man/man1/design_coupler.1
.TH design_coupler 1 "atlc-4.2.5 18/2/03" "Dr. David Kirkby"
.ds n 5
.SH NAME
\fBdesign_coupler\fR - for designing directional couplers (part of the \fBatlc\fR package)
.SH SYNOPSIS
\fBdesign_coupler [-C][-d][-e][-H height][-L length][-q]
.br
[s fstep][-Z Zo] CF fmin fmax \fR
.br
.SH DESCRIPTION
\fBdesign_coupler\fR is used to design directional couplers. It it \fBnot\fR
used to analyse couplers for which you know the dimensions. Instead, it
is used but when you require a coupler to have specific properties, but don't know the
required odd and even mode impedances or the required physical dimensions
that will achieve those required properties.
.P
As a minimum the user must specify the coupling factor CF in dB, the
minimum frequency fmin in MHz and the maximum frequency fmax in MHz.
With this information, the \fBdesign_coupler\fR will 
.br
a) Tell you the required odd and even mode impedances Zodd and Zeven
assuming the coupler is for 50 Ohms and assuming the coupler is is a 
quarter wave long, which might be an impractical length. There a numerous
ways of making a coupler having those impedances and \fBdesign_coupler\fR does
not (without the addition of options mentioned later), tell you how to
make such a coupler. 
b) Given you the frequency response of the coupler, making the
assumptions about the 50 Ohm impedance and quarter-wave length. The frequency
response is calculated at 5 points in the range specified by fmin and
fmax.
.P
By use of the -Z 'Zo' and -L 'length' and -f 'fstep' options it it posible to
specify different a different characteristic impedance, length and
different frequency steps to display the frequency response. 
.P
The computed  values of Zodd and Zeven required are valid no matter how
the coupler is design physically. So no matter whether it's implemented
on a PCB, air spaced or whatever, the above impedances are correct and
the frequency response is correct.
.P
The -d option causes \fBdesign_coupler\fR to not only report the
required odd and even modem impedances but also the physical
dimensions of a coupler that achieves these properties! \fBCurrently\fR,
the only stucture for which it is possible to compute the physical
dimentions is two wide edge-coupled striplines between two wide plates like
this:

.P
-----------------------------------------------------  ^
.br
|                                                   |  |
.br                                                     
|                  Er                               |  |
.br                                                     
|                                                   |  |
.br                                                     
|            -----------       -----------          |  H
.br
|            <----w----><--s--><----w---->          |  |
.br       
|                                                   |  |
.br
|                                                   |  |
.br
|                                                   |  |
.br
-----------------------------------------------------  v
.br
<-------------------------W------------------------->
.br
.P
The width W must be much greater than the height of the coupler and 
generally it is assumed that this width will at least 2*w+s*5*H,
otherwise the calculations will be incorrect. In order to calculate
these dimenisions an analytical method is used, which is only valid if 
the width W is infinity, but should be resonably good assuming W is at
least 2*w+s+5*H. 
.P 
It is later intended to enable design coupler to use other structures,
which migth be more suitable for construction, such as microstrip
couplers on PCBs, but for now at least, it is only possible to compute
the physical dimensions of the coupler using the above stucture. For
strong coupling (less than 20 dB or so), the dimenions calculated
might be impractical, as the spacing s will be so small. However, for
weak coupling, the physcical dimensions are practical. 
.SH OPTIONS

\fB-C\fR
.br
print copyright, licencing and copying information. 
.br
\fB-d\fR
.br
Design a coupler, using two edgle-coupled stiplines inside a wide
4-sided rectangular enclosure.

\fB-e\fR
.br
Priont an example of how to use \fBdesign_coupler\fR 
.br
\fB-H height\fR
.br
Specify the height of the enclosure in some convenient unit. By default,
a height of 1 unit is assumed, but by use of this option it is possible
to specify any height you want. Since its the ratio of dimensions that
is important, not the absolute values, this just scales all the other
dimensions by the specified height. It is just a conveneince for the
user. 
.br
\fB-L length\fR
.br
Specifies the coupler length in metres. By default the coupler is
assumed to be a quarter-wave, but this allow any length you want. Don't
chose a length that is a multiple of a half-wave though, as this will
make it impossible to couple any power out. 
\fB-q\fR
.br
This is the 'quite' switch and causes \fBdesign_coupler\fR to print out
less information. One can use -qq to cause the even less output.
.br
\fB-s fstep\fR
Causes \fBdesign_couler\fR to print out the frequency response at
different steps from the default 5 values. fstep must be in MHz. The
default value of fstep is obviously (fmax-fman)/5.
.br
\fBZ Zo\fR
.br
Causes \fBdesign_coupler\fR to compute properties of an impedance Zo
(shecified in Ohms). The default value for Zo is 50 Ohms.

.SH EXAMPLES
Run \fBdesign_coupler\fR gives examples of its use. However, here are
those same examples.
.P
Here are a examples of how to use \fBdesign_coupler\fR
In the examples, the % sign is used in front of anything you must type
which is what you will probably see when using the csh or tcsh as a shell. It
would probably be a $ sign if using the sh or bash shell. 
.P
To find the odd and even mode impedances and frequency response of a 50 Ohm
coupler, covering 130 to 170 MHz, with a coupling coefficient of 30 dB:
.P
% design_coupler 30 130 170
.P
Note the frequency response is symmetrical about the centre frequency at 0.192 dB
below that wanted. You may wish to redesign this for a coupling coefficient of 
about 29.9 dB, so the maximum deviation from the ideal 30.0 dB never exceeds 0.1 dB
Note the length suggested is 0.5 m (nearly 20") is a quarter wave at
the centre frequency of 150 MHz. You might find this a bit too long, so
let's specify a length of 0.25 m.
.P
% design_coupler -L 0.25 30 130 170
.P
What you may notice is that while the coupling to the coupled port is exactly
30 dB below the input power at the centre frequency (150 MHz) it is 
no longer symmetrical about the centre frequency. Also, deviations from the
ideal 30 dB are now much larger, with a maximum error of 1.012 dB
Unlike the case when the length is the default quarter wave, there is not much
you can do about this, since the deviations occur in both directions.
.P
Now assume you are reasonably happy with the response when the length is 250 mm
but would like to see the response at every 2.5 MHz. This can be done using the
-s option to design_coupler.
.P
% design_coupler -L 0.25 -s 2.5 30 130 170
.P
Assuming the performance is acceptable, the dimensions of the coupler can
be determined by adding the -d option. This will design a coupler that must look
like the structure below. The two inner conductors, which are spaced equally
between the top and bottom edges of the outer conductor, must be very thin.
These are placed along the length of a box of width W, height H and of
a length L determined by the user, which in this case is 250 mm.
.P
-----------------------------------------------------  ^
.br
|                                                   |  |
.br                                                     
|                  Er                               |  |
.br                                                     
|                                                   |  |
.br                                                     
|            -----------       -----------          |  H
.br
|            <----w----><--s--><----w---->          |  |
.br       
|                                                   |  |
.br
|                                                   |  |
.br
|                                                   |  |
.br
-----------------------------------------------------  v
.br
<-------------------------W------------------------->
.br
.P
The program reports: H = 1.0, ; w = 1.44 ; s = 0.44
The height of the box H must be small compared to the length L, (perhaps no
more than 7% of the length), or 17.5 mm in this case, with a length of 250 mm,
otherwise fringing effects will be significant. The width of the structure W
should be as large as possible. The program suggests making this 5*H+2*w+s. The
7% and 5*H+2*w+s are educated guesses, rather than exact figures. There is
no problem in making the width  larger than 5*H+2*w+s. The length L must be
kept at 250 mm. The RATIO of the dimensions H, w and s (but not L or W
must be kept constant. W just needs to be sufficiently large - it is
uncritical.  
.P
If you happened to have some 15 mm square brass available, then using that
for the side-walls would require that H becomes 15*1.0 = 15 mm,
w = 15*1.44 = 21.6 mm  and s = 15*0.44 = 6.6 mm
.P
There is no need to compute the above scaling with a calculator, as using
The -H option allows one to specify the height H. The program then reports the
exact dimensions for the length L, height H, w, s and suggests a minimum width
for W.
.P
In summary we have:
.br
    30 dB coupler +1.02 dB / -0.78 dB for 130 to 170 MHz
.br
    Length L = 250 mm, height H = 15 mm, stripline spacing s = 6.3 mm
.br
     stripline width w = 21.6 mm enclosure width W >= 124 mm
.P
By default, design_coupler prints a lot of information to the screen.
This can be reduced by the -q option or reduced to only one line with -qq
Other options include -Z to change the impedance from the default 50 Ohms
and -C to see the fully copyright, Licensing and distribution information
.br
.SH FILES
No files are created at all. 
.SH SEE ALSO
atlc(1)
.br
create_bmp_for_circ_in_circ(1)
create_bmp_for_circ_in_rect(1)
create_bmp_for_microstrip_coupler(1)
create_bmp_for_rect_cen_in_rect(1)
create_bmp_for_rect_cen_in_rect_coupler(1)
create_bmp_for_rect_in_circ(1)
create_bmp_for_rect_in_rect(1)
create_bmp_for_stripline_coupler(1)
create_bmp_for_symmetrical_stripline(1)
find_optimal_dimensions_for_microstrip_coupler(1)
.br
readbin(1)
.P 
.br
http://atlc.sourceforge.net                - Home page 
.br
http://sourceforge.net/projects/atlc       - Download area
.br
atlc-X.Y.Z/docs/html-docs/index.html       - HTML docs
.br
atlc-X.Y.Z/docs/qex-december-1996/design_coupler.pdf - theory paper
.br
atlc-X.Y.Z/examples                        - examples
.br
http://www.david-kirkby.co.uk              - my home page
.br
http://www.david-kirkby.co.uk/ham          - ham radio pages
.br


NEWS
NEWS for realease 4.3.0 on 23 May 2003.

A bug, preventing atlc from runningt properly on multiple
processors has been fixed. This is the main change. 

atlc should no longer fail any test on a multi-processor
system configured with --with-threads. The algorithm used
for both single processor and multiple processor (i.e. 
threaded code is the same). 

The -t option to atlc, which sets the number of threads 
when configured for multiple processors can be set to 0
to use entirely the single-threaded algorithm. If set to
one, it will use the multi-threaded algoritm, but use
only one thread. If set to some other number, it will use
that number of threads and be optimal for the same number
of cpus. 

Support for distributed programming has been temporarily removed.
Someone else added the code, but since he is no longer involved in
the development, it is difficult for me to follow. So I'll remove
the MPI specific code and will re-implement this. 

A test has been added to make sure atlc can handle the -d option 


Dr. David Kirkby 23/5/2003



src/non_gui/design_coupler.c
 /* atlc - arbitrary transmission line calculator, for the analysis of
transmission lines are directional couplers. 

Copyright (C) 2002. Dr. David Kirkby, PhD (G8WRB).

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either package_version 2
of the License, or (at your option) any later package_version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
USA.

Dr. David Kirkby, e-mail drkirkby@ntlworld.com 

*/

#include "definitions.h"

#ifdef HAVE_STDIO_H
#include <stdio.h>
#endif

#ifdef HAVE_STDLIB_H
#include <stdlib.h>
#endif

#ifdef SYS_TYPES_H
#include <sys/types.h>
#endif

#ifdef HAVE_MATH_H
#include <math.h>
#endif

#ifdef HAVE_STRINGS_H
#include <strings.h>
#endif

#ifdef HAVE_STRING_H
#include <string.h>
#endif

#ifdef HAVE_TIME_H
#include <time.h>
#endif

#ifdef HAVE_GSL_GSL_SF_ELLINT_H
#include <gsl/gsl_sf_ellint.h>
#endif

extern int errno;
int verbose=FALSE;

/* desgin_coupler does two very different things in the one program
1) Given a frequency range, the required coupling factor, it calcuates
the odd and even mode impedances needed for a coupler. It does this
assuming the length of the coupler is lambda/4, although you can vary
that on the command line with the -l option. 

2) Once the optimal values for the even and odd mode impedances are
found, it itteratively looks up the odd and even mode impedances for two
think lines of various widths (w) and spacings (s), looking for the
combination that gives the best rms error between the required
impedances and those that will result with the coupler design as
presented. 

It is assumed by default that the height of the box is of one unit (1
mm, 1" etc, but this may be changed on the command line. It only scales
the paramters w and s. 

*/

int main(int argc, char **argv) /* Read parameters from command line */
{
  int q, Hflag=FALSE;
  int calculate_physical_dimensions=FALSE;
  double er, Zo=-1, length=-1, fmin, fmax, fmean, fstep=-1, cf,  Zodd, Zeven; 
  double f, vcf, vcf_for_quarter_wave_line, w, s, error, error_max=1e30;
  double wanted_coupling_factor_in_dB, step=0.02, fq;
  double Zo_x=-1, Zeven_x=-1, Zodd_x=-1, best_s=-1, best_w=-1, height_of_box=1.0;
  double best_Zodd=-1, best_Zeven=-1, best_Zo=-1;
  while((q=get_options(argc,argv,"dCL:s:Z:H:")) != -1)
  switch (q) 
  {
    case 'd':
    calculate_physical_dimensions=TRUE;
    break;
    case 'C':
    print_copyright((char *) "2002");
    exit(1);
    break;
    case 'L':
    length=atof(my_optarg);
    break;
    case 'H':
    height_of_box=atof(my_optarg);
    Hflag=TRUE;
    break;
    case 's':
    fstep=atof(my_optarg);
    break;
    case 'Z':
    Zo=atof(my_optarg);
    break;
    case '?':
      usage_design_coupler();
    break;
  } /* End of the switch statement */

  if(argc-my_optind != 3)  /* This should be so hopefully !! */
  {
    usage_design_coupler();
    exit(1);
  }
  wanted_coupling_factor_in_dB=atof(argv[my_optind]);
  fmin=atof(argv[my_optind+1]);
  fmax=atof(argv[my_optind+2]);
  fmean=(fmin+fmax)/2.0;
  if(fstep <0 )
    fstep=(fmax-fmin)/4.0;
  if (wanted_coupling_factor_in_dB < 0.0 )
  {
    fprintf(stderr,"Enter the coupling factor as a positive number\n");
    exit(1);
  } 
  if (fmax <= fmin)
  {
    fprintf(stderr,"fmax <= fmin\n");
    exit(2);
  }
  if (Zo < 0.0)
    Zo=50.0; 
  if(length<0.0)
    length=75.0/fmean;  /* By default, make it a quarter wave long */
  /* The following sent in an email by Paul AA1L, sums the theory up 
  You make Zo=50=sqrt(Zoo*Zoe) and
  c=(Zoe-Zoo)/(Zoe+Zoo), c being the voltage coupling coefficient.
  I.e., for a 20dB coupler c=0.1 is the midband
  coupling.  
  Coupling varies as sin^2(f/fq), fq being frequency where the coupled
  length is a quarter wave.  

  HOWEVER, the above is not quite the full story, as that says coupling
  peaks at sin(1), when in fact its sin(Pi/2)
  */

  /* I need to find values for Zodd and Zeven to use, but first convert
  the coupling factor on the command line into the voltage coupling
  factor c */
  
  /* vfc stands for 'voltage coupling factor' */

  /* When the line is a quarter wave, one can get any amount of coupling
  you want, including a vfc of 1, in which case all the power transfers
  to the coupled port. Normally, the vcf will be less than 1.0. for a 20
  dB couplier is it 0.1 */

  vcf_for_quarter_wave_line=1.0/pow(10,wanted_coupling_factor_in_dB/20.0); 

  fq=75/length; /* frequency at which line is a quarter wave long */

  /* If the line is less than a quarter wave long, then less power is
  coupled, so to compensate we need to increase the voltage coupling
  factor 'vcf above that of the value for a quarter wave line.
  Since the 'vcf' varies as sin(0.5 *PI * f/fq)^2, where 
  fq is the frequency at which the line is a quarter-wave long, we must
  divide the vcf_for_quarter_wave_line by sin(0.5 *PI*f/fq)^2 to get
  the required vcf. */

  vcf=vcf_for_quarter_wave_line/pow(sin(0.5*M_PI*fmean/fq),2.0);
  
  /* Check that the voltage coupling factor does not exceed one */
  if ( vcf > 1.0 )
  {
    fprintf(stderr,"\n*****ERROR****\n");
    fprintf(stderr,"Sorry, you can't make a %6.3f dB coupler with a coupled line of %7.4f m long.\n",wanted_coupling_factor_in_dB, length);
    fprintf(stderr,"Either couple off a smaller fraction of the main power to the coupled port,\n");
    fprintf(stderr,"or make the line closer to an odd multiple of a quarter wave.\n");
    fprintf(stderr,"Odd mulitples of a quarter wave are: %.4f, %.4f, %.4f, %.4f .. m\n", 75/fmean, 3*75/fmean, 5*75/fmean, 7*75/fmean);
    exit(1);
  }

  /* After mucking around with Mathematica a bit, I found it was
  possible to invert the equations */
  
  Zodd = sqrt(1-vcf)*Zo/sqrt(1+vcf);
  Zeven=Zo*Zo/Zodd;

  printf("\nFor a %.3f dB %.3f Ohm coupler with a length of %.4f m,\n",wanted_coupling_factor_in_dB, Zo, length);
  printf("you need to have an odd-mode impedance Zodd to be %.3f Ohms and \n",Zodd);
  printf("an even mode impedance Zeven to be %.3f Ohms\n\n",Zeven);
  printf("%7.3f dB down <-- ************************** ---> %3.3f Ohm termination\n\n",wanted_coupling_factor_in_dB,Zo);
  printf("Drive this port --> ************************** ---> %3.3f Ohm termination\n",Zo);
  printf("                    <------- %8.4f m ----->\n",length);
  printf("\nDrive Port 1, coupler out of port 2 and terminate the other ports in Zo\n");
  printf("Such a coupler will have the response indicated below.\n\n");
  //printf("length =%.4f mean=%.3f vcf=%.3f vcf_for_quarter_wave_line=%.3f \n",length, fmean, vcf, vcf_for_quarter_wave_line);
  for(f=fmin; f<=fmax; f+=fstep)
  {
    cf=20*log10(vcf*pow(sin(0.5*M_PI*f/fq),2.0) ); /* This is what is now needed for some given length (and so fq) */
    printf("frequency = %.3f MHz coupling is %.3f dB down on the main arm\n",f,cf);
  }
  printf("\nYou may force the length to be any value you want using the -L option - it does\nnot have to be %.4f metres long\n",length);
  if(calculate_physical_dimensions==FALSE)
  {
    printf("You may try to find a coupler with these dimensions using the -d option\n\n");
    printf("Currently the -d option is not that fast, as it uses a brain-dead algorithm\n");
    printf("Hopefully one day the algorithm will be speeded up.\n");
  }
  if(calculate_physical_dimensions==TRUE)
  {
    er=1.0;
    printf("Please be patient - this will take a few minutes or so\n");
    for(s = 0.02; s<=100; s+=step)
    {
      for(w = 0.02; w<= 11.0; w += step)
      {
	/* Results are calculated assuming the box is one unit (mm, inch
	etc) high and later scaled */
        calculate_Zodd_and_Zeven(&Zodd_x, &Zeven_x, &Zo_x, w, 1.0, s, er);
	error=pow(Zodd-Zodd_x,2.0) + pow(Zeven-Zeven_x,2.0);
	if( error < error_max )
	{
	  best_s=s; 
	  best_w=w; 
	  best_Zo=sqrt(best_Zo * best_Zeven);
	  best_Zodd=Zodd;
	  best_Zeven=Zeven;
	  error_max=error;
        }
      }
    }
    //printf("w = %.4f s = %.4f which gives Zo = %.4f Zodd = %.4f Zeven = %.4f\n",best_w, best_s, best_Zo, best_Zodd, best_Zeven);
    /* Now try to get closer */
    /*
    for(s = best_s-step; s<=best_s+step; s+=step/1000)
    {
      for(w = best_w-step; w<= best_w+step; w += step/1000)
      {
        calculate_Zodd_and_Zeven(&Zodd_x, &Zeven_x, &Zo_x, w, 1.0, s, er);
	error=fabs(Zodd-Zodd_x) + fabs(Zeven-Zeven_x);
	if( error < error_max )
	{
	  best_s=s; 
	  best_w=w; 
	  best_Zodd=Zodd;
	  best_Zeven=Zeven;
	  error_max=error;
        }
      }
    }
    */
    best_Zo=sqrt(best_Zodd * best_Zeven);

    printf("|-----------^------------------------------------------------------------------|\n");
    printf("|           |                                                                  |\n");
    printf("|           |              <---w---><-----s----><---w-->                       |\n");
    printf("|           H              ---------            --------                       |\n");
    printf("|           |                                                                  |\n");
    printf("|           |   Er=1.0 (air)                                                   |\n");
    printf("------------v------------------------------------------------------------------\n");
    printf("<-----------------------------------------W----------------------------------->\n");
    printf("H =%.4f w = %.4f s = %.4f\n",height_of_box, height_of_box*best_w, height_of_box*best_s);
    printf("W must be *at least* %.4f, but larger does not matter.\n",5*height_of_box+ 2*best_w*height_of_box + height_of_box*best_s);
    printf("These dimensions give Zo = %.4f Zodd = %.4f Zeven = %.4f Ohms\n", best_Zo, best_Zodd, best_Zeven);
    if(Hflag==FALSE)
    {
      printf("****NOTE ****\n");
      printf("Although H is shown as 1.0, it can be 1 mm, 1 cm or even 1 m. It is important\n");
      printf("that w is %.4f times whatever H is, and that s is %.4f times whatever H is, but the absolute numbers are irrelavant.\n",best_w, best_s);
      printf("If you know the height H of your enclosure, use the -H option to indicate\n");
      printf("its value. This will ensure all the dimensions are scaled automatically for you.\n"); 
    }
    printf("****NOTE 2****\n");
    printf("The length *must* be %.4f m if you use these dimensions for W, H, w and s.\n",length); 
    printf("If %.4f m is inconvenient, change it with the -L option and recalculate\n to get new values of W, H, w and s\n",length);
    printf("See: http://atlc.sourceforge.net\n");
    printf("See: http://atlc.sourceforge.net/couplers.html\n");
  }
  exit(0);
}
